### ISSUE TITLE
[NuGet.org Bug]: Using nuget.core 2.14 with https://www.nuget.org/api/v2 no longer allows finding packages

### ISSUE THREAD

**Issue by EkiH** [NuGet/NuGetGallery#9182]
### Impact

I'm unable to use NuGet.org

### Describe the bug

We are using NuGet.Core 2.14.0 with https://www.nuget.org/api/v2 and yesterday we noticed an error blocking finding packages.

We are getting the following exception on almost all packages we try to find using FindPackages(...)
System.InvalidOperationException: The current value 'String' type is not compatible with the expected 'System.Int32' type.

This happens both in Azure and locally, no changes made on our part so we're suspecting this is due to some change on nuget.org.

Screenshot of a LINQPad script reproducing this:
![image](https://user-images.githubusercontent.com/6724459/181757455-1b009982-0978-4f0c-87be-d1d5a334f686.png)

### Repro Steps

1. Use Nuget.Core 2.14 NuGet package reference
2. Create a package repository targeting the v2 feed https://packages.nuget.org/api/v2
3. Try to find packages by id, e.g. Microsoft.Extensions.DependencyInjection.Abstractions
4. Get InvalidOperationException: The current value 'String' type is not compatible with the expected 'System.Int32' type.

Code:
`var repository = NuGet.PackageRepositoryFactory.Default.CreateRepository("https://packages.nuget.org/api/v2");`
`var packages = repository.FindPackagesById("Microsoft.Extensions.DependencyInjection.Abstractions"); // This throws: The current value 'String' type is not compatible with the expected 'System.Int32' type.`

### Expected Behavior

To get a list of packages like before

### Screenshots

![image](https://user-images.githubusercontent.com/6724459/181758324-81e003cd-e58e-425f-8d55-c95ef34af1b1.png)

### Additional Context and logs

This seems a bit similar to #9000 which we encountered in February

**Comment by EkiH** [comment]
@shishirx34 Should this be moved to the new sprint?

**Comment by shishirx34** [comment]
@joelverhagen - is this related to the #9000 issue? any thoughts?

**Comment by shishirx34** [comment]
@drewgillies - could you investigate this issue please?

**Comment by shishirx34** [comment]
@ryuyu  - to investigate.

**Comment by NeilTalbott** [comment]
For further context/reference, I see the same issue with (for example) NewtonSoft.Json - see LINQPad snippet below (again using NuGet 2.14):

```
void Main()
{
	var repository = NuGet.PackageRepositoryFactory.Default.CreateRepository("https://packages.nuget.org/api/v2");

	var version = new SemanticVersion("13.0.1");
	var package = repository.FindPackage("NewtonSoft.Json", version);

	Console.WriteLine(package.Id);
}
```

This change has had a wide-ranging impact for our software since our previously deployed bootstrapping installers no longer function correctly. Very grateful for any update on this.

**Comment by nickeeex** [comment]
This is still a pretty bad blocker and requires us to manually do some work to fetch the dependencies from the v3 feed and move them to the local cache.  Any progress on this issue would be highly appreciated.

**Comment by ryuyu** [comment]
Hey sorry for the slow response here. This appears to be related to an update we made to support download count values larger than the max value for int32. I'm digging into a solution here that should hopefully allow you to keep your workflow without changes.

**Comment by ryuyu** [comment]
I spoke with the package maintainers, and they said that the NuGet.Core package is no longer maintained and hasn't been for a while.
They recommended moving to the NuGet Client SDK packages, NuGet.Protocol and NuGet.Packaging.
The documentation for these packages can be found at https://docs.microsoft.com/en-us/nuget/reference/nuget-client-sdk
 
I believe the replacement you would want to use is documented in the Search section https://docs.microsoft.com/en-us/nuget/reference/nuget-client-sdk#search-packages
 
Please let me know if this doesn't work for your scenario, or if you have some difficulty migrating, and we can try to find another solution.

**Comment by nickeeex** [comment]
We have started the work to migrate to newer nuget libraries, we just currently have our own implementation for IPackage and IPackageLookup from NuGet.Core to use azure blobs as our package store. 

We could not find a similar way to do it with newer libraries. We are going to get rid of NuGet.Core but it will unfortunately still probably take a while to get there.

Bigger problem is that we have hundreds of deployments, some in azure and some locally at customers sites so updating them all is not a very simple task. 

If at all possible we would of course be happier with the v2 feed implementation being rolled back to a version where it does not break the NuGet.Core library.

**Comment by NeilTalbott** [comment]
We are in a similar situation to that described by nickeeex above. All of our customers are suddenly unable to use our bootstrapping mechanisms which have worked without problems over many years. Of course we will look into migrating to newer libraries but this will be a significant task, which itself will be substantially delayed because of the interim workarounds we will urgently need to provide to our clients (assuming we can find such workarounds).

In short, moving to the SDK packages doesn't work for us, at least not in the short/medium term. And (if I may say) making a catastrophic breaking change to the v2 feed, without warning, just to cope with package download counts over 2.1 billion seems a little crazy...

**Comment by martinczsmith** [comment]
Yes, I would echo the comments above. Our deployment activities have stopped dead too as our TeamCity server is essentially useless right now. I also fully agree with Neil, this was already infuriating but to then find out that this is due to a download counter......

**Comment by nickeeex** [comment]
> We are able to change the feed URL without too much of a hassle so for us that's one possible workaround, would still require a feed with all the packages from nuget.org.

We got a workaround for now. Mirrored all the packages with over 2.1 billion downloads to a myget.org feed that is used before nuget.org v2 feed in dependency resolution.  Only small problem was MyGet.org import from nuget.org was also broken for these packages with over 2.1 billion downloads 😅

**Comment by ryuyu** [comment]
Thank you for your patience on this matter.
We have been working on a fix that should unblock your scenario.
We are testing it now and it should be rolling out soon.
 
The root of the issue here is that the v2 endpoint was returning a value that could not be parsed into a valid int32 value.
Unfortunately, it looks like this was going to happen even without any changes as this value ends up coming from the search service, which has historically supported int64 download count values; we only just recently had packages that ended up exceeding the int32 max value.
Our fix will cap the download count values returned from the v2 endpoint to the max value of int32.
This should unblock the listed scenario. However, this does mean that seeing a download count of int32 max value from this endpoint now means that the download count is at least int32 max and does not speak to the actual value.
I still recommend continuing the effort to move to v3 as this will have all the real values, but hopefully this will give you the necessary runway.

**Comment by shishirx34** [comment]
The server side fix (`DownloadCount` capped at `Int32.MaxValue`) is now rolled out. I verified with the above repro steps, and the packages are being fetched as expected. Please let us know if you encounter any issues.
